lessons/07-mcp/README.md

```diff
@@ -0,0 +1,482 @@
+
+- The `@modelcontextprotocol/sdk` library provides the necessary tools to create and manage MCP servers.
+- `zod` is a TypeScript-first schema declaration and validation library that helps you define and validate the data structures used in your MCP server. 
+
+```bash
+npm install @modelcontextprotocol/sdk zod
+```
+
+Here's a simple example on how to create an MCP server using these libraries:
+
+```typescript
+import { McpServer, ResourceTemplate } from "@modelcontextprotocol/sdk/server/mcp.js";
+import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
+import { z } from "zod";
+ 
+// Create an MCP server
+const server = new McpServer({
+  name: "Demo",
+  version: "1.0.0"
+});
+```
+
+The preceding code does the following:
+
+- Imports the necessary classes and functions from the `@modelcontextprotocol/sdk` library.
+- Creates a server instance. The `McpServer` class is used to create the server, and it takes an object with the server's name and version as parameters.
+
+**You**: "This doesn't do much right?"
+
+**Time beetle**: "No, not yet. But we can add resources and tools to it. A tool is something that can be used to perform a specific task, while a resource is something that can be used to provide data or functionality to the tool. Let's add some below:
+
+```typescript
+server.tool("add",
+   { a: z.number(), b: z.number() },
+   async ({ a, b }) => ({
+     content: [{ type: "text", text: String(a + b) }]
+   })
+ );
+```
+
+In the preceding code, we:
+
+- Defined a tool called "add" that takes two numbers as input and returns their sum as output.
+- Described the input parameters and its validation using zod. The `z.number()` function is used to validate the input data, ensuring that both `a` and `b` are numbers.
+- The `async` function is used to perform the addition and return the result in a specific format.
+
+**You**: "Interesting, so the tool is like a function that can be called by the MCP client?"
+
+**Time beetle**: "Exactly! The MCP client can call this tool and pass the required parameters. Let's add a resource to the server as well:
+
+```typescript
+server.resource(
+   "greeting",
+   new ResourceTemplate("greeting://{name}", { list: undefined }),
+   async (uri, { name }) => ({
+     contents: [{
+       uri: uri.href,
+       text: `Hello, ${name}!`
+     }]
+   })
+ );
+```
+
+In this code, we:
+
+- Defined a resource called "greeting" that takes a name as input and returns a greeting message.
+- Used a resource template to define how to call the resource `greeting//{name}`. This schema is used to define the resource's URI format, which includes a placeholder for the name.
+- The `async` function is used to generate the greeting message based on the provided name.
+
+**You**: "So the resource is like a data source, this could be a database, file or even an API? If this was a for a file I would use a file:// URI?" like so:
```

Comment:
Do we want a comma before "like so". Otherwise the quote ends and then like so follows without any punctuation.